include/flang/Lower/PFTBuilder.h

```diff
@@ -115,54 +116,81 @@ static constexpr bool isFunctionLike{common::HasMember<
                   parser::SeparateModuleSubprogram>>};
 
 using LabelSet = llvm::SmallSet<parser::Label, 5>;
-using SymbolLabelMap = std::map<semantics::Symbol *, LabelSet>;
-
-/// Function-like units contain lists of evaluations.  These can be simple
-/// statements or constructs, where a construct contains its own evaluations.
-struct Evaluation {
-  using EvaluationTuple = common::CombineTuples<ActionStmts, OtherStmts,
-                                                ConstructStmts, Constructs>;
-
-  /// Hide non-nullable pointers to the parse-tree node.
-  /// Build type std::variant<const A* const, const B* const, ...>
-  /// from EvaluationTuple type (std::tuple<A, B, ...>).
-  template <typename A>
-  using MakeRefType = const A *const;
-  using EvaluationVariant = common::CombineVariants<
-      common::MapTemplate<MakeRefType, EvaluationTuple>>;
-  template <typename A>
-  constexpr auto visit(A visitor) const {
-    return std::visit(common::visitors{
-                          [&](const auto *p) { return visitor(*p); },
-                          [&](auto &r) { return visitor(r); },
-                      },
-                      u);
+using SymbolLabelMap = llvm::DenseMap<const semantics::Symbol *, LabelSet>;
+
+/// Provide a variant like container that can holds constant references.
+/// It is used in the other classes to provide union of const references
+/// to parse-tree nodes.
+template <typename... A>
+class ReferenceVariant {
+public:
+  template <typename B>
+  using ConstRef = common::Reference<const B>;
+
+  ReferenceVariant() = delete;
+  template <typename B>
+  ReferenceVariant(const B &b) : u{ConstRef<B>{b}} {};
+
+  template <typename B>
+  constexpr const B &get() const {
+    return std::get<ConstRef<B>> > (u).get();
   }
-  template <typename A>
-  constexpr const A *getIf() const {
-    auto *ptr = std::get_if<MakeRefType<A>>(&u);
-    return ptr ? *ptr : nullptr;
+  template <typename B>
+  constexpr const B *getIf() const {
```

Comment:
Since we're embracing the pointer paralysis full-on, should this be a pointer or some flavor of `std::optional<PointerFakingLifeAsAReferenceWrapperDuJour<B>>`?

Reply:
Yiik, that'd be interesting, but I'll pass.